core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java

```diff
@@ -394,12 +418,22 @@ public boolean process()
                 return false;
             }
 
-            // putIfAbsent will rehash automatically if rehash is needed, unless there isn't enough memory to do so.
-            // Therefore needRehash will not generally return true even if we have just crossed the capacity boundary.
-            while (lastPosition < positionCount && !needRehash()) {
-                // get the group for the current row
-                putIfAbsent(lastPosition, block);
-                lastPosition++;
+            int remainingPositions = positionCount - lastPosition;
+
+            long[] dummyGroupIds = new long[MAX_BATCH_SIZE];
+            while (remainingPositions != 0) {
+                int positionCountUntilRehash = maxFill - nextGroupId;
+                if (positionCountUntilRehash == 0) {
+                    if (!tryRehash()) {
```

Comment:
what if we ran out of memory inside a batch? e.g. the batch contains 64 new values but we only have space for 1?

Reply:
What do you mean run out of memory? Only rehash increases memory footprint and it returns false if there is not enough.

Reply:
so what happens if `tryRehash` returns false when called from `addNewGroup` inside a `batchedPutIfAbsent` call? is this handled somehow?

Reply:
It won't, because the batch size is not bigger than `positionCountUntilRehash`.
Which makes the `tryRehash` in `addGroup` a dead code.
Glad you noticed

Reply:
I missed the min with `positionCountUntilRehash` below.
It looks like in the unfortunate case when we have `maxFill - 1` groups, the batchSize will be 1 after a certain point. We could resize so that always `nextGroupId + min(remainingPositions, MAX_BATCH_SIZE) < maxFill`

Reply:
I was afraid of it at the beginning. But it appears that statistics is on our side. The case that you described requires number of groups to grow to some certain number and stop growing for a longer time. As this is probable with small overall number of groups, it is unlikely when then number of groups is big. The first rehash occurs at group number 12288, which means the chance of group count not rising any more is slim.